llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -16528,6 +16531,34 @@ static SDValue performUADDVAddCombine(SDValue A, SelectionDAG &DAG) {
     return DAG.getNode(Opcode, SDLoc(A), VT, Ext0.getOperand(0));
   };
 
+  // We can convert a UADDV(add(zext(64-bit source), zext(64-bit source))) into
+  // UADDLV(concat), where the concat represents the 64-bit zext sources.
+  auto DetectZextConcat = [&](SDValue A, SelectionDAG &DAG) {
+    // Look for add(zext(64-bit source), zext(64-bit source)), returning
+    // UADDLV(concat(zext, zext)) if found.
+    if (A.getOpcode() != ISD::ADD)
+      return SDValue();
+    EVT VT = A.getValueType();
+    if (VT != MVT::v4i32)
```

Comment:
Can this apply to other typesizes too? Maybe v8i16, and v2i64 if that one applies?
You might need to be careful with the return type from a v8i16 as with the way the nodes work at the moment it will be a v4i32 that probably needs to be NVCAST back to the correct type.

Reply:
Looks like there is only one other backend pattern currently for:

```

  def : Pat<(v2i64 (addlv (v4i32 V128:$Rn))),
            (v2i64 (SUBREG_TO_REG (i64 0), (!cast<Instruction>(Opc#"v4i32v") V128:$Rn), dsub))>;
```

Reply:
Some others were added recently, but have an v4i32 return type due to the way the intrinsics are specified.